SMILES property (#392)
Materials-Consortia/OPTIMADE: optimade.rst

  - A filter that matches an exactly given formula is :filter:`chemical_formula_anonymous="A2B"`.

> smiles

Reviewer: ```suggestion
chemical_formula_smiles
```

I think we preferred descriptive field names so far - do you think chemical formula is a valid prefix? Maybe structural_formula_smiles instead?

Author: This is a valid point. Prefix `structural_formula_` is more appropriate here, but I am not sure whether `formula` really fits. In the future we may wish to add InChI, InChIKey or DataWarrior ID (which is a descriptor, not an external ID) and to me they all seem constituting a group of similar properties. Maybe `chemical_descriptor_smiles` sounds better, reserving `chemical_descriptor_` prefix for other similar properties?

Reviewer: Sounds good - I wonder if this is another natural "namespace" we could support `_cheminformatics_` (or something catchier)... as soon as we add one of these descriptors to the top-level it will be diffiicult to justify excluding the others in the future. This is a bigger discussion, of course!

Author: I do not intend to use prefix-as-namespace for this property. We did not do so until now (`chemical_formula_` is just used to group similar properties), and sole namespace suggestion until now, `_cif_` (#329), was proposed to import a whole body of foreign properties. I do not think these few chemical descriptors are worth that, and I am not at all against adding properties for InChI, InChIKey.

Reviewer: Fair enough, I guess my rationale for suggesting a namespace is that the majority of entries served by OPTIMADE will *never* have a defined SMILES representation (as they are inorganics). I think the namespace discussion also involves  a more general issue of how new fields are treated in OPTIMADE; if we add `structural_formula_smiles` (or wahtever is decided) and release OPTIMADE v1.2, then the way the current spec is phrased, every implementation needs to *know* about that field so that they can return `null` if someone requests it, rather than returning an error. I think this is quite important to resolve before adding many more new fields. You could probably even argue that adding a field in OPTIMADE is currently a breaking change that should increment the major version number...

Author: I understand the problem with all providers needing to know all the OPTIMADE-supported properties, but I do not think that namespacing is the solution. The machine-readable list of all supported properties is provided for each OPTIMADE release (OpenAPI schemas) and could be quite easily read in by the implementations.

The point about property additions causing major version bump is strictly speaking fair, but I did not see a software/specification which would treat additions as API-breaking changes. Thus IMO OPTIMADE should not do this too. But we may surely bring this discussion up in the next meeting.

Author: @ml-evs I believe new consensus has emerged re this during 2022 workshop (was it yesterday?) that including new properties does not break backwards compatibility, or did I get it wrong?

Reviewer 2: So is there any final consensus on the field name? Mentioned options as well as several new ones:

- `smiles`
- `chemical_formula_smiles`
- `structural_formula_smiles`
- `chemical_smiles` (proposed by @BobHanson in the discussion of #368)
- `chemical_structure_smiles` (new addition)
- `chemical_connectivity_smiles` (new addition)
- `molecular_structure_smiles` (new addition)

I omitted the `chemical_descriptor_smiles` option since smiles is not really a descriptor in the context of cheminformatics.

Reviewer: > @ml-evs I believe new consensus has emerged re this during 2022 workshop (was it yesterday?) that including new properties does not break backwards compatibility, or did I get it wrong?

Sorry for missing this in the deluge @merkys. I think this was the conclusion, yes. New fields do not break client compatibility, though requesting e.g., `smiles` from an older implementation that does not implement it should raise an error (according the spec) where the client may be expecting `null` instead.

I no longer have any strong opinions on the naming of the field, given Andrius' arguments above against namespacing, and agree that the suggested prefixes feel a little bit unnatural.

Reviewer 3: @ml-evs 

> requesting e.g., smiles from an older implementation that does not implement it should raise an error (according the spec) where the client may be expecting null instead.

This "sounds" as potentially breaking clients, but IMO strictly speaking it is not. A well-implemented client for OPTIMADE version, say, v1.9.6 will always be able to know that `smiles` wasn't around until OPTIMADE v1.6.0, and thus can make sure not to break when an OPTIMADE v1.5.2 server returns an error when asked for `smiles`. My interpretation of semantic versioning is that it is when we introduce features that break *older* clients that we must increase the primary version number. 

But, that said, it may be a good point to clarify in the specification that clients should generally be prepared to handle servers returning errors for missing fields if they are at a lower version than the client.

Reviewer: I think that was a better summary of the consensus than mine :)

Author: @ml-evs @rartino @vaitkus Can we mark this discussion as resolved?

Reviewer 2: @merkys So we have settled on `smiles` for the field name?

Reviewer 2: @merkys `smiles` looks good to me too, so this conversation is resolved from my point of view.
Do  @ml-evs or @rartino have any objections to the field name?

Reviewer 3: I'm happy with `smiles`. (And I see I actually thumbed up @merkys suggestion to go with this a few comments up.)

Author: Oh right, I did miss @rartino's thumb up. I am going to assume @ml-evs is also fine with `smiles` as per his comment:
> I no longer have any strong opinions on the naming of the field, given Andrius' arguments above against namespacing, and agree that the suggested prefixes feel a little bit unnatural.

~~~~~~

> - **Description**: The SMILES descriptor for the structure.

Reviewer: ```suggestion
- **Description**: The SMILES (Simplified Molecular Input Line Entry Specification) descriptor for the structure.
```

- **Type**: string
- **Requirements/Conventions**:

  - **Support**: OPTIONAL support in implementations, i.e., MAY be :val:`null`.
  - **Query**: Support for queries on this property is OPTIONAL.

>     Queries MUST treat the value of this property as a string, not the represented chemical structure.

Reviewer: ```suggestion
    Queries MUST treat the value of this property as a raw string, without SMILES-specific semantics.
```

>     Queries MUST treat the value of this property as a raw string, without SMILES-specific semantics.
>     That is, providers MUST NOT perform substructure search, just regular string comparison.

Reviewer 4: ```suggestion
```
A molecule can have hundreds of valid SMILES descriptors. A client would have to include all of them in a query, to determine whether a particular molecule is present in the database.
I can imagine that such a query would be slow to execute.
A more efficient way, would be to convert the SMILES string of the query into a structure and then  back into a SMILES string using the same method that was used to generate the SMILES strings in the database. 
These lines however explicitly forbid databases from implementing this method.

Author: I [have responded to these concerns already](https://github.com/Materials-Consortia/OPTIMADE/issues/368#issuecomment-986231183).

Reviewer 2: @JPBergsma, are you OK with leaving these two lines intact and marking the conversation as resolved?

From what I understand from the discussions in #392, it was agreed to implement the complex structure search functionality in a different way (e.g. by using SMARTS).

  - MUST adhere to the `OpenSMILES specification v1.0 <http://opensmiles.org/opensmiles.html>`__.
  - When structures or their parts cannot be unambiguously depicted in SMILES according to OpenSMILES recommendations, using the guidelines from `Quirós et al. 2018 <https://jcheminf.biomedcentral.com/articles/10.1186/s13321-018-0279-6>`__ is RECOMMENDED.

>   - Providers MAY canonicalize produced SMILES descriptors, but this is not mandatory.

Reviewer: It's not clear to me (as a non-expert) what canonicalize means in this context - could you add a few words describing it?

Author: Good point - I will clarify it.

Reviewer 4: ```suggestion
  - Providers are RECOMMANDED to canonicalize produced SMILES descriptors, but this is not mandatory.
```
I would recommand using a canonicalization method so that if the client knows which canonicalization method has been used, it no longer has to generate hundreds of different SMILES structures for a query.

Author: See [my comment on #368](https://github.com/Materials-Consortia/OPTIMADE/issues/368#issuecomment-986231183) why I did not defined a common canonicalization method.

Author: @ml-evs does 72a5409 make it clearer, or should I elaborate it more?

Reviewer: That's perfect for me! I will resolve.

  - A filter that matches an exactly given formula is :filter:`chemical_formula_anonymous="A2B"`.

smiles
~~~~~~

> - **Description**: The SMILES (Simplified Molecular Input Line Entry Specification) descriptor for the structure.

Reviewer 2: The term "descriptor" should be avoided since it has a specific value in cheminformatics.

- **Type**: string
- **Requirements/Conventions**:

  - **Support**: OPTIONAL support in implementations, i.e., MAY be :val:`null`.
  - **Query**: Support for queries on this property is OPTIONAL.
    Queries MUST treat the value of this property as a raw string, without SMILES-specific semantics.
    That is, providers MUST NOT perform substructure search, just regular string comparison.
  - MUST adhere to the `OpenSMILES specification v1.0 <http://opensmiles.org/opensmiles.html>`__.
  - When structures or their parts cannot be unambiguously depicted in SMILES according to OpenSMILES recommendations, using the guidelines from `Quirós et al. 2018 <https://jcheminf.biomedcentral.com/articles/10.1186/s13321-018-0279-6>`__ is RECOMMENDED.

>   - Providers MAY canonicalize (i.e., use rules to establish stable order of atoms) produced SMILES descriptors, but this is not mandatory.

Reviewer 2: The term "descriptor" should be avoided since it has a specific value in cheminformatics.

  - A filter that matches an exactly given formula is :filter:`chemical_formula_anonymous="A2B"`.

smiles
~~~~~~

> - **Description**: The SMILES (Simplified Molecular Input Line Entry System) representation of the structure.

Reviewer: I think we need a bit more clarification of the expected use.

How "much" of the structure should be described by the SMILES string for it to be valid here (e.g., that it should appear in the results when someone searches for it?) Do we need to require that every "site" in the OPTIMADE structure is present in the SMILES string? Obviously for `nperiodic_dimensions=0` and a single molecule this makes sense, same for an `nperiodic_dimensions=3` molecular crystal, but what about:

- co-crystal with two distinct molecules (does SMILES do something fancy for this already?)
- an inorganic surface with adsorbed molecule
- a hybrid perovskite structure with molecular unit as a cation

Author: > How "much" of the structure should be described by the SMILES string for it to be valid here (e.g., that it should appear in the results when someone searches for it?) Do we need to require that every "site" in the OPTIMADE structure is present in the SMILES string?

This is a good point. I would say that every "site" has to be represented in SMILES. There surely will be situations where this is not attainable (i.e., OpenSMILES cannot express polymers and there will be difficulties in depicting mixture sites). Maybe at this point it would be easier to say that only the structures that are "expressible" using OpenSMILES should have `smiles`, that is, no nonstandard approximations should be done.

> Obviously for `nperiodic_dimensions=0` and a single molecule this makes sense, same for an `nperiodic_dimensions=3` molecular crystal, but what about:
> 
> * co-crystal with two distinct molecules (does SMILES do something fancy for this already?)

SMILES can contain many distinct molecules, disconnected components are joined with `.` (if I get the question right)

> * an inorganic surface with adsorbed molecule
> * a hybrid perovskite structure with molecular unit as a cation

I would say these two fall under class "polymer", thus inexpressible.

Reviewer 5: > co-crystal with two distinct molecules (does SMILES do something fancy for this already?
This can be described using a "dot bond", e.g. CuSO4.O.O

> - **Description**: The SMILES (Simplified Molecular Input Line Entry Specification) representation of the structure.

Reviewer 2: ```suggestion
- **Description**: The SMILES (Simplified Molecular Input Line Entry System) representation of the structure.
```

- **Type**: string
- **Requirements/Conventions**:

  - **Support**: OPTIONAL support in implementations, i.e., MAY be :val:`null`.
  - **Query**: Support for queries on this property is OPTIONAL.
    Queries MUST treat the value of this property as a raw string, without SMILES-specific semantics.
    That is, providers MUST NOT perform substructure search, just regular string comparison.

>   - MUST adhere to the `OpenSMILES specification v1.0 <http://opensmiles.org/opensmiles.html>`__.

Reviewer 2: ```suggestion
  - Value MUST adhere to the `OpenSMILES specification v1.0 <http://opensmiles.org/opensmiles.html>`__.
```

>   - When structures or their parts cannot be unambiguously represented in SMILES according to OpenSMILES recommendations, using the guidelines from `Quirós et al. 2018 <https://jcheminf.biomedcentral.com/articles/10.1186/s13321-018-0279-6>`__ is RECOMMENDED.

Reviewer 2: ```suggestion
  - When structures or their parts cannot be unambiguously represented in SMILES according to OpenSMILES recommendations, using the guidelines from `Quirós et al. 2018 <https://doi.org/10.1186/s13321-018-0279-6>`__ is RECOMMENDED.
```

  - Providers MAY canonicalize (i.e., use rules to establish stable order of atoms) produced SMILES representations, but this is not mandatory.

>     Generally, providers SHOULD NOT change the representation more frequently than the structure itself is modified.

Reviewer: Can we provide a couple of examples for people who are unfamiliar with SMILES (without needing them to click out of the spec and read the paper/OpenSMILES spec)? Below is taken from wikipedia, so please check if the string is actually OpenSMILES compliant...

```suggestion
    Generally, providers SHOULD NOT change the representation more frequently than the structure itself is modified.
    
  - **Examples**:
      - caffeine: `CN1C=NC2=C1C(=O)N(C(=O)N2C)C`
```
